src/query/src/window_sort.rs

```diff
@@ -0,0 +1,3101 @@
+    {
+        match $data_type {
+            arrow_schema::DataType::Timestamp(arrow_schema::TimeUnit::Second, _) => {
+                $m!(TimestampSecondType $(, $args)*)
+            }
+            arrow_schema::DataType::Timestamp(arrow_schema::TimeUnit::Millisecond, _) => {
+                $m!(TimestampMillisecondType $(, $args)*)
+            }
+            arrow_schema::DataType::Timestamp(arrow_schema::TimeUnit::Microsecond, _) => {
+                $m!(TimestampMicrosecondType $(, $args)*)
+            }
+            arrow_schema::DataType::Timestamp(arrow_schema::TimeUnit::Nanosecond, _) => {
+                $m!(TimestampNanosecondType $(, $args)*)
+            }
+            $($p => $fallback,)*
+        }
+    };
+}
+
+/// Find the slice(where start <= data < end and sort by `sort_column.options`) from the given range
+///
+/// Return the offset and length of the slice
+fn find_slice_from_range(
+    sort_column: &SortColumn,
+    range: &TimeRange,
+) -> datafusion_common::Result<(usize, usize)> {
+    let ty = sort_column.values.data_type();
+    let time_unit = if let DataType::Timestamp(unit, _) = ty {
+        unit
+    } else {
+        return Err(DataFusionError::Internal(format!(
+            "Unsupported sort column type: {}",
+            sort_column.values.data_type()
+        )));
+    };
+    let array = &sort_column.values;
+    let opt = &sort_column.options.unwrap_or_default();
+    let descending = opt.descending;
+
+    let typed_sorted_range = [range.start, range.end]
+        .iter()
+        .map(|t| {
+            t.convert_to(time_unit.into())
+                .ok_or_else(|| {
+                    DataFusionError::Internal(format!(
+                        "Failed to convert timestamp from {:?} to {:?}",
+                        t.unit(),
+                        time_unit
+                    ))
+                })
+                .and_then(|typed_ts| {
+                    let value = Value::Timestamp(typed_ts);
+                    value
+                        .try_to_scalar_value(&value.data_type())
+                        .map_err(|e| DataFusionError::External(Box::new(e) as _))
+                })
+        })
+        .try_collect::<_, Vec<_>, _>()?;
+
+    let (min_val, max_val) = (typed_sorted_range[0].clone(), typed_sorted_range[1].clone());
+
+    // get slice that in which all data that `min_val<=data<max_val`
+    let (start, end) = if descending {
+        // note that `data < max_val`
+        // i,e, for max_val = 4, array = [5,3,2] should be start=1
+        // max_val = 4, array = [5, 4, 3, 2] should be start= 2
+        let start = bisect::<false>(&[array.clone()], &[max_val.clone()], &[*opt])?;
+        // min_val = 1, array = [3, 2, 1, 0], end = 2
+        // min_val = 1, array = [3, 2, 0], end = 2
+        let end = bisect::<false>(&[array.clone()], &[min_val.clone()], &[*opt])?;
+        (start, end)
```

Comment:
if `min_val = 1, array = [3, 2, 1, 0], end = 2`, we won't take value `1` as end is exlusive.

Reply:
> if `min_val = 1, array = [3, 2, 1, 0], end = 2`, we won't take value `1` as end is exlusive.

min_val is inclusive as it's get the mininal value from PartitionRange